refactor: Simplify whiteboard access management with new multi-user flags

gustavotrott opened the pull request:
It greatly simplifies how the system manages who can draw on the whiteboard.

Two new flags have been introduced:

* `user.whiteboardWriteAccess`: Determines whether a user can draw on the whiteboard. This flag remains unchanged when switching slides or presentations.
* `meeting.multiUserWhiteboardEnabled`: Determines whether users automatically receive `whiteboardWriteAccess` upon joining.

The whiteboard button now toggles the `meeting.multiUserWhiteboardEnabled` flag.

For breakout rooms, a new configuration option is available:
`breakoutRoomsMultiUserWhiteboardDefaultOn=true`: defines whether the room starts with `meeting.multiUserWhiteboardEnabled` enabled.

For the main room, a new `/create` parameter has been added:
`multiUserWhiteboardEnabled`: allows enabling multi-user whiteboard mode by default when creating a meeting.

Closes: #22893, #19860, #6582, #21098, #22787

**Future work:** Allow moderators to toggle `meeting.multiUserWhiteboardEnabled`, since the current toolbar button is visible only to the presenter.

gustavotrott committed:
introduce user.whiteboardWriteAccess and meeting_usersPolicies.multiUserWhiteboardEnabled

JoVictorNunes committed:
refactor(html5/voice-activity): Rework voice activity local state

gustavotrott committed:
middleware handle getUserMutedStateStream and getUserTalkingStateStream streams

JoVictorNunes committed:
Improve typescript

JoVictorNunes committed:
Fix typescript

JoVictorNunes committed:
Fix TS

gustavotrott committed:
Add user info to UserTalkingVoiceEvtMsgBody

JoVictorNunes committed:
Update talking state when user gets muted

gustavotrott committed:
create new redis msg UserVoiceStateEvtMsg

JoVictorNunes committed:
Remove redundant stream

gustavotrott committed:
add more situations to send UserVoiceStateEvtMsg

gustavotrott committed:
remove cache of userVoiceState on meeting end

JoVictorNunes committed:
Handle user exited audio

gustavotrott committed:
stop requiring user state for user voice status msg

gustavotrott committed:
remove comments

gustavotrott committed:
send userVoiceState when UserDisconnectedFromGlobalAudioMsgHdlr

gustavotrott committed:
introduce new field leftVoiceConf for user_voice_activity_stream

JoVictorNunes committed:
Fix talking indicator not disappearing when stopping talking

JoVictorNunes committed:
Fix condition

gustavotrott committed:
Merge remote-tracking branch 'upstream/v3.0.x-release' into whiteboard-access-as-user-prop

jfederico committed:
fix: bigbluebutton-tests/playwright/package.json & bigbluebutton-tests/playwright/package-lock.json to reduce vulnerabilities (#24116)

The following vulnerabilities are fixed with an upgrade:
- https://snyk.io/vuln/SNYK-JS-PLAYWRIGHTCORE-13553173

Co-authored-by: snyk-bot <snyk-bot@snyk.io>

antobinary committed:
build: Add dependency on libmagickcore-extra (bbb-web)

Seems for some servers libmagickcore-6.q16-6-extra is not auto installed and slides do not get properly converted https://github.com/bigbluebutton/bigbluebutton/issues/23726
Here I am trying to add the dependency to the -extra package in a more flexible way (not only for -6.q16-6)

antonbsa committed:
tests(fix): avoid race condition by waiting expected text content when pressing play/pause (#24198)

gustavotrott committed:
better manegment of user.whiteboardWriteAccess and meeting.multiUserWhiteboardEnabled

gustavotrott committed:
New config breakoutRoomsMultiUserWhiteboardDefaultOn=true

gustavotrott committed:
Add new /create param multiUserWhiteboardEnabled

gustavotrott committed:
fixes from review

gustavotrott commented on akka-bbb-apps/src/main/scala/org/bigbluebutton/core/apps/users/SetUserWhiteboardWriteAccessReqMsgHdlr.scala:
```diff
@@ -0,0 +1,143 @@
+          "pen_tool",
+          "app.whiteboard.toolbar.multiUserLimitHasBeenReachedNotification",
+          "Notification when the maximum number of whiteboard writers has been reached",
+          Map("numberOfUsers" -> maxNumberOfActiveUsers.toString)
+        )
+        outGW.send(notifyEvent)
+      }
+
+      for {
+        userId <- usersToSetWhiteboardWrite.take(availableSlots)
+        newUserState <- Users2x.setUserWhiteboardWriteAccess(liveMeeting.users2x, userId, msg.body.whiteboardWriteAccess)
+      } yield {
+        broadcast(newUserState)
+
+        if (newUserState.whiteboardWriteAccess) {
+          val notifyEvent = MsgBuilder.buildNotifyUserInMeetingEvtMsg(
+            newUserState.intId,
+            liveMeeting.props.meetingProp.intId,
+            "info",
+            "pen_tool",
+            "app.whiteboard.available",
+            "Notification to when the user received whiteboard access",
+            Map()
+          )
+          outGW.send(notifyEvent)
+          NotificationDAO.insert(notifyEvent)
+        } else {
+          val notifyEvent = MsgBuilder.buildNotifyUserInMeetingEvtMsg(
+            newUserState.intId,
+            liveMeeting.props.meetingProp.intId,
+            "info",
+            "pen_tool",
+            "app.whiteboard.unavailable",
+            "Notification to when the user lost whiteboard access",
+            Map()
+          )
+          outGW.send(notifyEvent)
+          NotificationDAO.insert(notifyEvent)
+        }
+      }
```
No, the presenter has a different validation where he can draw even when whiteboardWriteAccess is false.

gustavotrott committed:
Update docs/docs/development/api.md

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

Tainan404 committed:
Feat(html5): Move rtt-check fetch to a worker (#24152)

* Feat(html5): Move rtt-check fetch to a worker

* Fix typo on clear of timeout

gabriellpr committed:
test: pass the test upload multiple presentations to flaky

gabriellpr committed:
test: update lock viewers annotations test

antonbsa committed:
test(fix): increase sleep duration for presentation minimize and add checks for zoom buttons

gustavotrott committed:
bbb-web: Avoid error for requests expecting return in plain/text

gustavotrott committed:
gql-server: Improve database indexes

Tainan404 commented:
When the last user that is not the presenter lose whiteboard access the count on the presentation toolbar is zeroed and the multiwhitboard is considered disabled on current 3.0, on the new implementation the presenter is considered on the count and the count isn't zeroed
Left: 3.0
Right: PR implementation

[Screencast from 07-11-2025 19:15:08.webm](https://github.com/user-attachments/assets/ac66386d-7ac8-4b23-a46c-67d652cc2ee3)

gustavotrott commented:
> When the last user that is not the presenter lose whiteboard access the count on the presentation toolbar is zeroed and the multiwhitboard is considered disabled on current 3.0, on the new implementation the presenter is considered on the count and the count isn't zeroed Left: 3.0 Right: PR implementation
> 
>  Screencast.from.07-11-2025.19.15.08.webm

Yes, it is different by design.
Now the button is used to turn on/off the multi-user whiteboard. So it will be ON even without any user assigned. Because it will still be applied for joining users.
and the presenter is part of the count as he will continue with write access if someone else get the presenter. so the count indicate the number of users with the flag `whiteboardWriteAccess` and the presenter user is included as he has the flag (we just can't see because the presenter flag has priority).

gustavotrott committed:
remove isCurrentlyTyping from hasura metadata

gustavotrott committed:
fix undefined variable guestStatusVal for handleJoinExistingUser

transifex-integration[bot] committed:
Translate en.json in ja

100% translated source file: 'en.json'
on 'ja'.

Tainan404 committed:
Fix(html5): Make the share webcam removes away status (#24184)

* Fix(html5): Make the share webcam disable away status as join audio does

* Reimplents the away logic

antonbsa committed:
refactor(tests): Migrate automated tests files to TS + add lint formatting (ESLint with Prettier) (#24035)

* refactor(test): migrate files to Typescript

- Adjusted locator method calls
- Replaced `sleep` calls with `page.waitForTimeout`
- Ensured all imports from core elements and utilities are updated to reflect the new file structure
- Add husky with pre-commit command to validate TS

* style(tests): add ESLint and Prettier config and apply changes to all files

- auto-check on pre-commit with husky
- add docs section for formatting

* refactor(tests): migrate all test suite files to TS

- snapshots folder changes to contain "ts"

* chore: add eslint-plugin-simple-import-sort

* fix: update upload PDF wait time and close all previous notification on save settings test

* perf(tests): add SWC and improve tsconfig.json for faster TypeScript compilation

* test(fix): apply several suggestion by coderabbit

gustavotrott committed:
resolve conflict

ramonlsouza committed:
make breakout rooms inherit lock settings from parent meeting

ramonlsouza committed:
do not display lock viewers option in breakout rooms

ramonlsouza committed:
move pinned notes notification to backend

JoVictorNunes committed:
style(poll): Increase visible area of typed responses

transifex-integration[bot] committed:
Translate en.json in tr

100% translated source file: 'en.json'
on 'tr'.

transifex-integration[bot] committed:
Translate en.json in et

100% translated source file: 'en.json'
on 'et'.

JoVictorNunes committed:
fix(html5/user-list): Append both user name and user id

gustavotrott committed:
Merge branch 'v3.0.x-release' into whiteboard-access-as-user-prop

transifex-integration[bot] committed:
Translate en.json in ru

100% translated source file: 'en.json'
on 'ru'.

ramonlsouza committed:
only copy private chat lock settings on breakout creation

germanocaumo committed:
fix(tldraw): remove unnecessary extra zoom calls +

- Only reset zoom when fitTwoWdith actually changes, it was being set on every presentation mount
- Round viewed region up to 100 percent when its too near (avoids 99.99... values)

ramonlsouza committed:
adjust quiz correct answer validation to work with any language

transifex-integration[bot] committed:
Translate en.json in it_IT

100% translated source file: 'en.json'
on 'it_IT'.

prlanzarin committed:
fix: LiveKit token update fails on database

When a new LK token is generated for an existing session, its insertion
in the database is failing due to the usage of forceInsert. This is not
the expected behavior - the token should be updated.
While the behavior is mostly benign as token re-generation does not yet
play a significant role, this might bork unrelated batched DB ops and
should be fixed.

Fix the issue by using insertOrUpdate so that LK tokens are properly
updated.

transifex-integration[bot] committed:
Translate en.json in ar

100% translated source file: 'en.json'
on 'ar'.

antobinary committed:
chore: Bump release to 3.0.19

gabriellpr committed:
[3.0] Adds the test upload multiple presentation on ci (#24218)

* add test upload multiple presentation to run on ci

* changed util to receive two options for the notification

pinhaum committed:
(tests): Add tests for "create breakout rooms" buttons (#24183)

* test: add check to create breakout rooms button existence

* refac: update logic to fit typescript migration

* fix: ensure mangeUsers dropdown is closed after continuing

pinhaum committed:
(tests): Add tests for breakout rooms sharing audio and video (#24150)

* test: refac to check user joining breakout with audio shared in the main room

* tests: add tests for user joining breakout with audio and video in the main room

* tests: add tests for user sharing audio after joining breakout rooms

* tests: update to check with webcamContainer selector

* refac: update logic to fit typescript migration

* refac: remove duplicate checks from joinRoom

pinhaum committed:
(tests): Add test for "group parameters" (breakout room pre-selection) (#24061)

* test: add test for group parameters while create breakout rooms

* fix: code rabbit suggestions

* tests: update to check if correct user has joined the breakout room

Co-authored-by: Anton Barboza de Sá <antonbsa.bck@gmail.com>

* tests: Update to check room name in fullscreen

Co-authored-by: Anton Barboza de Sá <antonbsa.bck@gmail.com>

* tests: Update username check to compare with literal strings

Co-authored-by: Anton Barboza de Sá <antonbsa.bck@gmail.com>

* tests: refac roomNameInput

* refac: update logic to fit typescript migration

---------

Co-authored-by: Anton Barboza de Sá <antonbsa.bck@gmail.com>
